Add Weblate Fedora Messaging schema

harriebird opened the pull request:
## Proposed changes

This adds a new schema for Weblate Fedora Messaging.
This solves issue [fedora_messaging#268](https://github.com/WeblateOrg/fedora_messaging/issues/268).

## Checklist

- [ ] Lint and unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added documentation to describe my feature.
- [ ] I have squashed my commits into logic units.
- [ ] I have described the changes in the commit messages.

## Other information

nijel commented:
Looks good for the body schema, now we need to wrap it in Python classes (see https://fedora-messaging.readthedocs.io/en/stable/user-guide/messages.html). Maybe header schema will be needed as well (I didn't investigate that).

nijel commented on weblate_schemas/message.py:
```diff
@@ -0,0 +1,106 @@
+        return "https://weblate.org/static/weblate-128.png"
+
+
+class WeblateMessage(BaseMessage):
```
Probably should include versioning to allow future changes:

```suggestion
class WeblateV1Message(BaseMessage):
```

Zlopez replied:
You don't need to have it for the initial message schema, but will need that for any updated message schema. The first one could be `WeblateMessage`, but the changed one would need to be `WeblateV2Message`. Still a good idea to have the versioning in from start.

Keep in mind that you only need to do that if you are removing something from the message schema, adding things doesn't break anything.

nijel commented on weblate_schemas/message.py:
```diff
@@ -0,0 +1,106 @@
+        """Initialize the WeblateMessage class with loading of the body_schema."""
+        super().__init__(**kwargs)
+        self.body_schema = load_schema("weblate-messaging.schema.json")
+
```
At least `summary` should also be present, based on https://pagure.io/fedora-infrastructure/issue/8291#comment-857387.

Zlopez replied:
I would also recommend to add properties that are used by FMN as `usernames` and `packages`. See https://fedora-messaging.readthedocs.io/en/stable/user-guide/messages.html#example-schema

Also each Fedora application should have `app_name` property.

nijel reviewed (commented):
Can we add some kind of testing that the message classes work on the expected payload?

Zlopez reviewed (commented):
Few comments for the schema

harriebird commented:
Hi everyone, thank you for your comments. The latest commit contains the following changes:

- Renamed the message class to `WeblateV1Message`
- Added properties `summary` and `usernames` into the `WeblateV1Message` class
- Added entry point
- Replaced `jsonschema` with `jsonschema[format]` to extend format validation support
- Added new tests

nijel commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
+fedora-messaging
```
This should be an optional dependency, there is no need to require it for somebody wanting other schemas.

nijel commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
+fedora-messaging
 fqdn
-jsonschema
+jsonschema[format]
```
Perhaps this belongs to `requirements-test.txt`?

nijel commented on setup.py:
```diff
@@ -11,4 +11,13 @@
 
-setup(install_requires=REQUIRES)
+setup(
+    name="weblate_schemas",
```
Already present in `setup.cfg`.

nijel commented on setup.py:
```diff
@@ -11,4 +11,13 @@
+            "base.message=weblate_schemas.messages:BaseMessage",
+            "weblate.message=weblate_schemas.messages:WeblateV1Message",
+        ]
+    },
```
This should go to `setup.cfg`.

abompard commented on setup.py:
```diff
@@ -11,4 +11,13 @@
+    name="weblate_schemas",
+    entry_points={
+        "fedora.messages": [
+            "base.message=weblate_schemas.messages:BaseMessage",
```
Please prefix the key here with `weblate.`. The `base.message` key is already used by Fedora Messaging

abompard commented on setup.py:
```diff
@@ -11,4 +11,13 @@
+    entry_points={
+        "fedora.messages": [
+            "base.message=weblate_schemas.messages:BaseMessage",
+            "weblate.message=weblate_schemas.messages:WeblateV1Message",
```
I recommend suffixing this key with `.v1`, as it is the version 1 of the schema.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,117 @@
+
+    def __init__(self, **kwargs) -> None:
+        """Initialize BaseMessage class."""
+        super().__init__(**kwargs)
```
This block is not necessary, no need to redefine `__init__` if you're just calling the superclass.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,117 @@
+        return self.body["user"]
+
+    @property
+    def id(self) -> int:
```
The `id` property is reserved by Fedora Messaging. You can use something like `change_id` if you want.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,117 @@
+    def usernames(self):
+        """Return the usernames involved."""
+        users = {self.body["author"], self.body["user"]}
+        return list(users)
```
It would make testing easier if the list is sorted alphabetically, but this is just a non-blocking suggestion.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,117 @@
+
+    def __str__(self) -> str:
+        """Return a human-readable representation of the message."""
+        return f"{self.action} {self.timestamp}"
```
FYI, the string representation of the message is what is going to be in the email body when people ask to be notified of these messages in FMN. You may want to put a little more text there, but it's just a non-blocking suggestion.

nijel replied:
@harriebird Can we have this covered by the tests? It would be annoying if it would break.

@abompard `action` is a human-readable description of the action, so that looks okay. But I'm not sure whether to include the timestamp here.

abompard replied:
Yeah you can put the timestamp there, it's really up to you. I would set something more human readable though, or just make it return `self.summary` if you can't or don't want to have a longer text.

nijel commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,117 @@
+# SPDX-License-Identifier: MIT
+
+"""Message class implementation for Weblate Fedora Messaging."""
+
```
```suggestion

from __future__ import annotations

```

Without that you end up with `TypeError: unsupported operand type(s) for |: 'type' and 'types.GenericAlias'` on Python 3.9.

harriebird commented:
Hi everyone, I just updated the code based on your comments. Below are the changes made:

- Reverted changes in `setup.py` and moved the `entry_points` declaration to `setup.cfg`
- Updated the `summary` property implementation
- Updated the `usernames` property implementation
- Changed the `id` property into `change_id`
- Used `get` method in accessing the body dictionary
- Added `pip install -e .` command in the dependency install of `test.yml` workflow

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+    @property
+    def agent_name(self) -> str:
+        """The username who cause the action."""
+        return self.body.get("user", "")
```
If there is no agent name, it should be `None`, not an empty string

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+    @property
+    def change_id(self) -> int:
+        """Return the change ID."""
+        return self.body.get("change_id")
```
According to the JSON schema, `change_id` is required, you can be sure it'll always be there, so you can use `self.body["change_id"]`. (just a suggestion, `get()` will work too)

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+    @property
+    def action(self) -> str:
+        """Return the change verbose name."""
+        return self.body.get("action")
```
According to the JSON schema, `action` is required, you can be sure it'll always be there, so you can use `self.body["action"]`.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+    @property
+    def timestamp(self) -> datetime:
+        """Return the timestamp of the change."""
+        return self.body.get("timestamp")
```
According to the JSON schema, `timestamp` is required, you can be sure it'll always be there, so you can use `self.body["timestamp"]`.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+    def summary(self) -> str:
+        """Return the message summary."""
+        return "{} event{} occurred in {}.".format(
+            self.body.get("action"),
```
Same remark here. By the way, you can use `self.action`: if you end up moving the data in the dictionary, you'll only have to update it in one place.

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+            " done by {}".format(self.body.get("user"))
+            if self.body.get("user", None)
+            else "",
+            self.body.get("timestamp"),
```
Same remark here

abompard commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,126 @@
+        """Return the usernames involved."""
+        users = []
+        users += [self.author] if self.author else []
+        users += [self.user] if self.user else []
```
Or something like:
```py
users = list(sorted({name for name in (self.author, self.user) if name}))
```
(just a suggestion, not a blocker)

It is useful to convert to a list in the end, because `sorted()` returns an iterator, not an actual list.

abompard reviewed (approved):
LGTM! :+1:

nijel commented on weblate_schemas/messages.py:
```diff
@@ -0,0 +1,116 @@
+
+    def __str__(self) -> str:
+        """Return a human-readable representation of the message."""
+        return f"{self.action} {self.timestamp}"
```
```suggestion
        return f"{self.summary}"
```

nijel commented:
I've scheduled this for merge because I want to release this package soon anyway. In case we hit some issues later, those can be fixed in the next release.
